docs/event_listeners.md

```diff
@@ -0,0 +1,252 @@
+# Event Listeners
+
+When building a blueprint, your application may require to listen to events. Events can be of any type, and handling those events is entirely up to your discretion.
+
+In general, when defining your job, you must register any event listeners and provide a context as such:
+
+```rust
+#[job(
+    id = 0,
+    params(x),
+    result(_),
+    event_listener(TangleEventListener, MyEventListener1, MyEventListener2, ...), // <-- Register all event listeners here
+    verifier(evm = "MyVerifier")
+)]
+pub fn hello_event_listener(
+    x: u64,
+    context: MyContext, // <-- The context type must be the first additional parameter
+    env: GadgetConfiguration<parking_lot::RawRwLock>,
+) -> Result<u64, Infallible> {
+    Ok(x.saturating_pow(2u32))
+}
+```
+
+In order to make these registered event listeners to work, we must define structs that implement `EventListener`.
+## Creating Event Listeners
+
+To create an event listener, begin by defining a struct or enum which will listen to events:
+
+```rust
+use gadget_sdk::event_listener::EventListener;
+use async_trait::async_trait;
+
+/// Define a simple event listener that ticks every MS milliseconds.
+pub struct Ticker<const MSEC: usize> {
+    additional_delay: u64,
+}
+```
+
+Then, define a context for the event listener:
+
+```rust
+#[derive(Copy, Clone)]
+pub struct MyContext {
+    pub additional_delay: u64,
+}
+```
+
+Next, implement `EventListener` for `Ticker`:
+
+```rust
+/// Implement the [`EventListener`] trait for the Ticker struct. [`EventListener`] has two type parameters:
+/// 
+/// - T: is the type of the event that the listener listens for. This can be anything that is Send + Sync + 'static.
+/// In this case, we are using [`Instant`], which is a timestamp.
+///
+/// - Ctx: is the context type that the listener receives when constructed. This can be anything that is Send + Sync + 'static,
```

Comment:
I feel this can be provided differently, even though I can mentally map `Ctx` to `MyContext` (because they're similar). I don't think it's connected super well though. This amount of inline comments should maybe be motivated outside of the code block.